src/gt4py/next/iterator/type_system/type_synthesizer.py

```diff
@@ -116,15 +116,17 @@ def _(lhs: ts.ScalarType, rhs: ts.ScalarType) -> ts.ScalarType | ts.TupleType:
 
 
 @_register_builtin_type_synthesizer
-def deref(it: it_ts.IteratorType) -> ts.DataType:
+def deref(it: it_ts.IteratorType) -> ts.DataType | ts.DeferredType:
```

Comment:
Just a question: is the type annotation for `it` correct? Shouldn't it be a union with `DeferredType`? If that is the case, why is not type checker complaining here?
```suggestion
def deref(it: it_ts.IteratorType| ts.DeferredType) -> ts.DataType | ts.DeferredType:
```